db/migrate/20200728054557_refer_config_proxy_to_http_proxy.rb

```diff
@@ -0,0 +1,60 @@
+class ReferConfigProxyToHttpProxy < ActiveRecord::Migration[6.0]
+  def up
+    add_column :foreman_virt_who_configure_configs, :http_proxy_id, :integer
+
+    if User.where(login: User::ANONYMOUS_ADMIN).exists?
+      User.as_anonymous_admin do
+        ::ForemanVirtWhoConfigure::Config.find_each do |config|
+          if config.proxy.present?
+            create_and_refer_http_proxy_if_not_exist(config)
+          end
+        end
+      end
+    end
+
+    remove_column :foreman_virt_who_configure_configs, :proxy
+  end
+
+  def down
+    add_column :foreman_virt_who_configure_configs, :proxy, :string
+
+    if User.where(login: User::ANONYMOUS_ADMIN).exists?
+      User.as_anonymous_admin do
+        ::ForemanVirtWhoConfigure::Config.find_each do |config|
+          if config.http_proxy_id.present?
+            config.proxy = config.http_proxy.url
+            config.save!
+          end
+        end
+      end
+    end
+
+    remove_column :foreman_virt_who_configure_configs, :http_proxy_id
+  end
+
+  private
+
+  def create_and_refer_http_proxy_if_not_exist(config)
+    http_proxy = ::HttpProxy.find_by(url: proxy_url(config.proxy))
+
+    if http_proxy
+      config.http_proxy_id = http_proxy.id
+    else
+      http_proxy = HttpProxy.new(name: "virt_who_#{config.proxy}", url: proxy_url(config.proxy))
+
+      if http_proxy.save
+        config.http_proxy_id = http_proxy.id
+      end
+    end
+
+    config.save!
+  end
+
+  def proxy_url(proxy)
```

Comment:
There's possibly a better way to do this, using `URI`.

Look at these examples:

```
[2] pry(main)> uri = URI('192.168.1.1')
=> #<URI::Generic 192.168.1.1>
[3] pry(main)> uri.scheme
=> nil
```

```
[4] pry(main)> uri_two = URI('http://example.com:8080')
=> #<URI::HTTP http://example.com:8080>
[5] pry(main)> uri_two.scheme
=> "http"
[6] pry(main)> uri_two.to_s
=> "http://example.com:8080"
```
```
[8] pry(main)> uri_three = URI('https://admin:changeme@192.168.1.1')
=> #<URI::HTTPS https://admin:changeme@192.168.1.1>
[9] pry(main)> uri_three.scheme
=> "https"
[11] pry(main)> uri_three.user
=> "admin"
[12] pry(main)> uri_three.password
=> "changeme"
[13] pry(main)> uri_three.to_s
=> "https://admin:changeme@192.168.1.1"
```

`HttpProxy` has an instance method called `full_url` which might be a more accurate check, especially since the current text field on ::Config has pretty much no validation. `full_url` takes into account any username+password as well as the url+port

```
[15] pry(main)> HttpProxy.last.full_url
=> "http://admin:changeme@example.com:8080"
```

`full_url` is not saved in the DB - it's calculated on the instance after it's loaded. If we go this route, it may be easier to pass a list of all HttpProxies to `create_and_refer_http_proxy_if_not_exist`. That means we can load all of them at once and avoid a find_by for each VirtWhoConfig

TLDR: The method could probably be something more like this:

```
def proxy_uri(config_proxy)
  uri = URI (config_proxy) # may need to handle errors here?
  unless uri.scheme
    uri.scheme = 'http'
   end
   uri
end
```

Check it out - I think there could be an advantage. If there are difficulties then let me know